File: Java Project/Iteration_2/src/RegistrationSystem.java
package Iteration_2.src;


import org.apache.log4j.Logger;

import java.util.ArrayList;
import java.util.Objects;

public class RegistrationSystem {
    static org.apache.log4j.Logger logger = Logger.getLogger(RegistrationSystem.class.getName());

    public void checkCourseIsTakenBefore(EnrollmentRequest enrollmentRequest) {

        Student student = enrollmentRequest.getStudent();
        ArrayList<CompulsoryCourse> courses = enrollmentRequest.getCourses();
        ArrayList<CompulsoryCourse> takenCourses = new ArrayList<CompulsoryCourse>();
        logger.info("System Checking " + student.getStudentID() + "'s enrollment requests");

        for (CompulsoryCourse course : courses) {
            if (student.getTranscriptBefore().getTakenCourses().stream().map(CompulsoryCourse::getCourseCode).toList().contains(course.getCourseCode())) {
                System.out.println("Course " + course.getCourseCode() + " can not be taken because it has already been taken.");
                logger.warn("Course " + course.getCourseCode() + " failed 'taken before test'. ");
                //Log
            } else {
                logger.info("Course " + course.getCourseCode() + " passed 'taken before test'. ");
                takenCourses.add(course);
                //Log

            }


        }
        enrollmentRequest.setCourses(takenCourses);

    }

    public void checkPrerequisites(EnrollmentRequest enrollmentRequest) {
        ArrayList<CompulsoryCourse> compulsaryCourses = enrollmentRequest.getCourses();
        Student student = enrollmentRequest.getStudent();
        ArrayList<CompulsoryCourse> takenCourses = new ArrayList<CompulsoryCourse>();
        for (CompulsoryCourse course : compulsaryCourses) {
            if (course.getPrerequisites() != null) {
                if (!student.getTranscriptBefore().getTakenCourses().stream().map(CompulsoryCourse::getCourseCode).toList().containsAll(course.getPrerequisites())) {
                    System.out.println("Course " + course.getCourseCode() + " can not taken because prerequisites are not satisfied.");
                    logger.info("Course " + course.getCourseCode() + " failed 'prerequisites ' test");
                } else {
                    logger.info("Course " + course.getCourseCode() + " passed 'prerequisites' test ");
                    takenCourses.add(course);
                }


                //Log
            } else {
                System.out.println("Course " + course.getCourseCode() + "  passed 'prerequisites' test ");
                logger.info("Course " + course.getCourseCode() + " passed 'prerequisites' test ");
                takenCourses.add(course);
            }


        }
        enrollmentRequest.setCourses(takenCourses);

    }

    public void isOpenedThisTerm(EnrollmentRequest enrollmentRequest, String term) {
        ArrayList<CompulsoryCourse> courses = enrollmentRequest.getCourses();
        Student student = enrollmentRequest.getStudent();
        ArrayList<CompulsoryCourse> takenCourses = new ArrayList<CompulsoryCourse>();
        for (CompulsoryCourse course : courses) {
            if (Objects.equals(course.getOpenedTerm(), term)) {
                takenCourses.add(course);
                logger.info("Course " + course.getCourseCode() + " passed 'opened this term ' test");
            } else {
                System.out.println("Course " + course.getCourseCode() + " can not been taken because it does not opened this term");
                logger.info("Course " + course.getCourseCode() + " failed 'opened this term ' test");
            }
        }
        enrollmentRequest.setCourses(takenCourses);

    }


}

File: Python/src/RegistrationSystem.py
import logging

class RegistrationSystem:
    def __init__(self):
        self.__logger = logging.getLogger(__name__)

    def check_course_is_taken_before(self, enrollment_request):
        student = enrollment_request.student
        courses = enrollment_request.courses
        taken_courses = []
        self.__logger.info(f"System Checking {student.student_id}'s enrollment requests")


        for course in courses:
            if course.course_code in [c.course_code for c in student.transcript_before.taken_courses]:
                print(f"Course {course.course_code} can not be taken because it has already been taken.")
                self.__logger.warning(f"Course {course.course_code} failed 'taken before' test.")
            else:
                self.__logger.info(f"Course {course.course_code} passed 'taken before' test.")
                taken_courses.append(course)

        enrollment_request.courses = taken_courses

    def check_prerequisites(self, enrollment_request):
        courses = enrollment_request.courses
        student = enrollment_request.student
        taken_courses = []

        for course in courses:
            if course.prerequisites is not None:
                if not all(prereq in [c.course_code for c in student.transcript_before.taken_courses] for prereq in
                           course.prerequisites):
                    print(f"Course {course.course_code} can not be taken because prerequisites are not satisfied.")
                    self.__logger.info(f"Course {course.course_code} failed 'prerequisites' test.")
                else:
                    self.__logger.info(f"Course {course.course_code} passed 'prerequisites' test.")
                    taken_courses.append(course)
            else:
                print(f"Course {course.course_code} passed 'prerequisites' test.")
                self.__logger.info(f"Course {course.course_code} passed 'prerequisites' test.")
                taken_courses.append(course)

        enrollment_request.courses = taken_courses

    def is_opened_this_term(self, enrollment_request, term):
        courses = enrollment_request.courses
        student = enrollment_request.student
        taken_courses = []

        for course in courses:
            if course.opened_term == term:
                taken_courses.append(course)
                self.__logger.info(f"Course {course.course_code} passed 'opened this term' test.")
            else:
                print(f"Course {course.course_code} can not been taken because it does not opened this term")
                self.__logger.info(f"Course {course.course_code} failed 'opened this term' test.")

        enrollment_request.courses = taken_courses
